File: agent/crates/public/src/leaky_bucket.rs
/*
 * Copyright (c) 2022 Yunshan Networks
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::{
    sync::{
        atomic::{AtomicBool, AtomicU64, Ordering},
        Arc,
    },
    thread::{self, JoinHandle},
    time::Duration,
};

const TICK_INTERVAL: Duration = Duration::from_millis(100);
const TICK_PER_SECOND: u64 =
    (Duration::from_secs(1).as_millis() / TICK_INTERVAL.as_millis()) as u64;
const BURST_MULTIPLE: u64 = 10;

pub struct LeakyBucket {
    rate: Arc<AtomicU64>,
    token: Arc<AtomicU64>,
    running: Arc<AtomicBool>,

    handle: JoinHandle<()>,
}

impl LeakyBucket {
    pub fn new(rate: Option<u64>) -> Self {
        let running = Arc::new(AtomicBool::new(true));
        let rate = Arc::new(AtomicU64::new(rate.unwrap_or(0)));
        let token = Arc::new(AtomicU64::new(0));

        let t_running = running.clone();
        let t_rate = rate.clone();
        let t_token = token.clone();
        let handle = thread::Builder::new()
            .name("leaky-bucket".to_owned())
            .spawn(move || {
                let mut rate = 0;
                let mut quantity_per_tick = 0;
                let mut full = 0;
                let token = t_token;
                while t_running.load(Ordering::Relaxed) {
                    let new_rate = t_rate.load(Ordering::Relaxed);
                    if new_rate == 0 || new_rate != rate {
                        rate = new_rate;
                        if rate == 0 {
                            thread::park();
                            continue;
                        }
                        quantity_per_tick = 1.max(rate / TICK_PER_SECOND);
                        full = quantity_per_tick * BURST_MULTIPLE;
                        token.store(full, Ordering::Release);
                    }

                    let _ = token.fetch_update(Ordering::Release, Ordering::Relaxed, |t| {
                        if t + quantity_per_tick > full {
                            None
                        } else {
                            Some(t + quantity_per_tick)
                        }
                    });

                    thread::park_timeout(TICK_INTERVAL);
                }
            })
            .unwrap();

        LeakyBucket {
            rate,
            token,
            running,
            handle,
        }
    }

    pub fn set_rate(&self, rate: Option<u64>) {
        self.rate.store(rate.unwrap_or(0), Ordering::Relaxed);
        self.handle.thread().unpark();
    }

    pub fn acquire(&self, size: u64) -> bool {
        if self.rate.load(Ordering::Relaxed) == 0 {
            return true;
        }

        self.token
            .fetch_update(Ordering::Release, Ordering::Relaxed, |t| {
                if t < size {
                    None
                } else {
                    Some(t - size)
                }
            })
            .is_ok()
    }
}

impl Default for LeakyBucket {
    fn default() -> Self {
        Self::new(None)
    }
}

impl Drop for LeakyBucket {
    fn drop(&mut self) {
        self.running.store(false, Ordering::Relaxed);
        self.handle.thread().unpark();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rate() {
        for (i, rate) in vec![1000, 10000000].into_iter().enumerate() {
            let bucket = if i % 2 == 0 {
                let bucket = LeakyBucket::new(None);
                bucket.set_rate(Some(rate));
                bucket
            } else {
                LeakyBucket::new(Some(rate))
            };
            thread::sleep(TICK_INTERVAL / 10);
            assert!(
                bucket.acquire(BURST_MULTIPLE * rate / TICK_PER_SECOND),
                "failed acquiring for rate {}",
                rate
            );
            assert!(!bucket.acquire(1), "failed leaking for rate {}", rate);
            thread::sleep(TICK_INTERVAL + TICK_INTERVAL / 10);
            for _ in 0..10 {
                assert!(
                    bucket.acquire(rate / 10 / TICK_PER_SECOND),
                    "failed acquiring for rate {}",
                    rate
                );
            }
            assert!(
                !bucket.acquire(rate / 10 / TICK_PER_SECOND),
                "failed leaking for rate {}",
                rate
            );
        }

        let unlimited = LeakyBucket::new(None);
        for _ in 0..1000 {
            assert!(
                unlimited.acquire(u64::MAX),
                "failed acquiring for unlimited rate"
            );
        }
    }
}

File: server/libs/datastructure/leaky_bucket.go
/*
 * Copyright (c) 2024 Yunshan Networks
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package datastructure

import (
	"math"
	"time"
)

type LeakyBucket struct {
	rate     uint64
	interval time.Duration // the unit is ms
	token    uint64
	full     uint64
	quantity uint64 // get token number per timer
	timer    *time.Ticker
	multiple uint64 // burst multiple
	running  bool
	spin     SpinLock
}

func (b *LeakyBucket) Init(rate uint64) {
	b.interval = 100
	b.multiple = 10
	// setup timer
	b.timer = time.NewTicker(time.Duration(b.interval) * time.Millisecond)

	// initial token, feed to the bucket full
	b.SetRate(rate)
	b.token = b.full
	b.running = true

	go b.run()
}

func (b *LeakyBucket) Close() {
	b.running = false
	b.timer.Stop()
}

func (b *LeakyBucket) run() {
	for b.running == true {
		<-b.timer.C

		// fill token
		b.spin.Lock()
		if b.token+b.quantity > b.full {
			b.token = b.full
		} else {
			b.token += b.quantity
		}
		b.spin.Unlock()
	}
}

func (b *LeakyBucket) SetRate(rate uint64) {
	if rate == 0 {
		rate = math.MaxUint64
	}
	quantity := rate / uint64(time.Second/time.Millisecond/b.interval)
	if quantity == 0 { // for 1 <= rate < 10
		quantity = 1
	}
	b.spin.Lock()
	b.rate = rate
	b.quantity = quantity
	b.full = b.quantity * b.multiple
	b.spin.Unlock()
}

func (b *LeakyBucket) Acquire(size uint64) bool {
	if b.rate == math.MaxUint64 {
		return true
	}

	b.spin.Lock()
	if b.token < size {
		b.spin.Unlock()
		return false
	}
	b.token -= size
	b.spin.Unlock()
	return true
}
